### PR TITLE
grib: replace DataSource and derived classes with VSILFILE directly.

### PR DESCRIPTION
**Pull request by gogglesguy**
DataSource and co served the same purpose as VSILFILE, ie. a wrapper to access both regular files and memory buffers in a file like manner. DataSource probably predates VSILFILE and its interface doesn't support 64 bit file offsets. As DataSource is currently a thin layer around VSILFILE, it seems only reasonable to simply remove this extra layer and get native 64 bit offset support.

## Tasklist
 - [ ] Review
 - [x] All CI builds and checks have passed

### CODE REVIEW

File: gdal/frmts/grib/gribdataset.cpp [hunk 1]
```diff
@@ -941,15 +936,28 @@ GDALDataset *GRIBDataset::Open( GDALOpenInfo *poOpenInfo )
...
-    MemoryDataSource mds(poOpenInfo->pabyHeader, poOpenInfo->nHeaderBytes);
-    if (ReadSECT0(mds, &buff, &buffLen, -1, sect0, &gribLen, &version) < 0) {
+
+    CPLString tmpFilename;
+    tmpFilename.Printf("/vsimem/gribdataset-%p", poOpenInfo);
+
+    VSILFILE * memfp = VSIFileFromMemBuffer(tmpFilename, poOpenInfo->pabyHeader, poOpenInfo->nHeaderBytes, FALSE);
+    if (memfp == nullptr || ReadSECT0(memfp, &buff, &buffLen, -1, sect0, &gribLen, &version) < 0) {
+
+        if (memfp != nullptr) {
```
Comment (rouault) [inline gdal/frmts/grib/gribdataset.cpp]: indentation looks a bit odd

Comment (gogglesguy) [reply gdal/frmts/grib/gribdataset.cpp]: Oops, should be fixed in next push.

**Comment by rouault** [comment]
Can you squash the commits in a single one ?

**Comment by gogglesguy** [comment]
Squashed + 1 minor indentation fix.

**Pull request merged**